src/client/client.ts

```diff
@@ -0,0 +1,219 @@
+import request from 'superagent';
+import * as utils from '../utils/utils';
+import IntDecoding from '../types/intDecoding';
+
+interface ErrorWithAdditionalInfo extends Error {
+  rawResponse: string | null;
+  statusCode: number;
+}
+
+function createJSONParser(options: utils.JSONOptions) {
+  return (
+    res: request.Response,
+    // eslint-disable-next-line no-unused-vars
+    fnOrStr: string | ((err: Error, obj: any) => void)
+    // eslint-disable-next-line consistent-return
+  ) => {
+    if (typeof fnOrStr === 'string') {
+      // in browser
+      return fnOrStr && utils.parseJSON(fnOrStr, options);
+    }
+
+    // in node
+    // based off https://github.com/visionmedia/superagent/blob/1277a880c32191e300b229e352e0633e421046c8/src/node/parsers/json.js
+    res.text = '';
+    res.setEncoding('utf8');
+    res.on('data', (chunk) => {
+      res.text += chunk;
+    });
+    res.on('end', () => {
+      let body: any;
+      let err: ErrorWithAdditionalInfo | null;
+      try {
+        body = res.text && utils.parseJSON(res.text, options);
+      } catch (err_) {
+        err = err_;
+        // issue #675: return the raw response if the response parsing fails
+        err.rawResponse = res.text || null;
+        // issue #876: return the http status code if the response parsing fails
+        err.statusCode = res.status;
+      } finally {
+        fnOrStr(err, body);
+      }
+    });
+  };
+}
+
+/**
+ * Remove falsy values or values with a length of 0 from an object.
+ * @param obj
+ */
+function removeFalsyOrEmpty(obj: Record<string, any>) {
+  for (const key in obj) {
+    if (Object.prototype.hasOwnProperty.call(obj, key)) {
+      // eslint-disable-next-line no-param-reassign
+      if (!obj[key] || obj[key].length === 0) delete obj[key];
+    }
+  }
+  return obj;
+}
+
+type Query<F> = {
+  format?: F;
+  [key: string]: any;
+};
+
+/**
+ * getAcceptFormat returns the correct Accept header depending on the
+ * requested format.
+ * @param query
+ */
+/* eslint-disable no-redeclare,no-unused-vars */
+function getAcceptFormat(
+  query?: Query<'msgpack' | 'json'>
+): 'application/msgpack' | 'application/json' {
+  /* eslint-enable no-redeclare,no-unused-vars */
+  if (
+    query !== undefined &&
+    Object.prototype.hasOwnProperty.call(query, 'format')
+  ) {
+    switch (query.format) {
+      case 'msgpack':
+        return 'application/msgpack';
+      case 'json':
+      default:
+        return 'application/json';
+    }
+  } else return 'application/json';
+}
+
+export interface TokenHeader {
+  'X-Algo-API-Token': string;
+}
+
+export interface KMDTokenHeader {
+  'X-KMD-API-Token': string;
+}
+
+export default class HTTPClient {
+  private address: string;
+  private tokenHeader: TokenHeader | KMDTokenHeader;
+  public intDecoding: IntDecoding = IntDecoding.DEFAULT;
+
+  constructor(
+    tokenHeader: string | TokenHeader | KMDTokenHeader,
```

Comment:
Indexer tokens use the `X-Indexer-API-Token` header, which is missing from here.